### PR TITLE
floating-point predictor added

### PR DESCRIPTION
**Pull request by TomasKralCZ**
This PR implements floaating-point predictor from https://github.com/image-rs/image-tiff/issues/89.

Only images with 32-bit samples are tested, as I don't have access to any 64-bit ones.

### CODE REVIEW

**Comment by fintelia** [comment]
This looks good from a quick skim. Sorry, I haven't gotten a chance to read up on TIFF predictors and go through this in detail.

**Comment by TomasKralCZ** [comment]
Oh, that's ok, this isn't urgent...

File: src/decoder/predictor.rs [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    for row in 0..height {
+        for col in samples..width * samples {
+            let prev_pixel = image[(row * img_width * samples + col - samples)];
+            let pixel = &mut image[(row * img_width * samples + col)];
+            *pixel = pixel.wrapping_add(prev_pixel);
+        }
+    }
+}
+
+pub fn fp_predict<T>(
```
Comment (fintelia) [inline src/decoder/predictor.rs:72]: I'm not sure it makes sense for `fp_predict` to be generic. If there was just a `fp_predict_f32` and `fp_predict_f64` they could both be much simpler. I think you should be able to do something along the lines of: 
```rust
fn fp_predict_f32(buf: &mut [f32], (width, height): (usize, usize), res_width: usize, samples: usize) {
    for row in 0..height {
        rev_hpredict_nsamp(AsNeMutBytes::as_ne_mut_bytes(
            &mut buf[row * res_width * simples_per_pixel..][..width * samples_per_pixel]), ...);
        for i in 0..(width*samples_per_pixel) {
             buf[i] = f32::from_bits(buf[i].to_bits().from_be())
        }
    }
}

**Review by fintelia — COMMENTED** [review]
Added some comments. I think the general structure could be significantly simplified, but the overall logic looks good to me.

File: src/decoder/predictor.rs (hunk 1 above)
Comment (aschampion) [reply src/decoder/predictor.rs:72]: FWIW there are 16- and 24-bit floating point TIFFs, the 16-bit version not being that rare in the wild so we might have an interest in supporting it at some point. Libtiff supports both.

**Comment by TomasKralCZ** [comment]
Sorry, I was busy with other things, but I'll try to finish this soon...

**Comment by fintelia** [comment]
https://github.com/image-rs/image-tiff/pull/165 is now merged. Thanks @TomasKralCZ for contributing the initial design!

**Pull request closed without merging**